src/commonMain/kotlin/exchange.dydx.abacus/calculator/TriggerOrdersInputCalculator.kt

```diff
@@ -1,10 +1,160 @@
 
+import abs
 import exchange.dydx.abacus.protocols.ParserProtocol
+import exchange.dydx.abacus.utils.Numeric
+import exchange.dydx.abacus.utils.mutable
+import exchange.dydx.abacus.utils.safeSet
 
 @Suppress("UNCHECKED_CAST")
 internal class TriggerOrdersInputCalculator(val parser: ParserProtocol) {
-    internal fun calculate(state: Map<String, Any>): Map<String, Any> {
-        return state
+    internal fun calculate(
+        state: Map<String, Any>,
+        subaccountNumber: Int?,
+    ): Map<String, Any> {
+        val account = parser.asNativeMap(state["account"])
+        val subaccount = if (subaccountNumber != null) {
+            parser.asNativeMap(
+                parser.value(
+                    account,
+                    "subaccounts.$subaccountNumber",
+                ),
+            )
+        } else {
+            null
+        }
+        val triggerOrders = parser.asNativeMap(state["triggerOrders"])
+        val marketId = parser.asString(triggerOrders?.get("marketId"))
+        val stopLossOrder = parser.asNativeMap(triggerOrders?.get("stopLossOrder"))
+        val takeProfitOrder = parser.asNativeMap(triggerOrders?.get("takeProfitOrder"))
+        val position = parser.asNativeMap(parser.value(subaccount, "openPositions.$marketId"))
+
+        return if (triggerOrders != null && position != null) {
+            val modified = state.mutable()
+            val modifiedStopLossOrder = if (stopLossOrder != null) {
+                calculateTriggerOrderTrade(stopLossOrder, position)
+            } else {
+                stopLossOrder
+            }
+            val modifiedTakeProfitOrder = if (takeProfitOrder != null) {
+                calculateTriggerOrderTrade(takeProfitOrder, position)
+            } else {
+                takeProfitOrder
+            }
+            val modifiedTriggerOrders = triggerOrders.mutable()
+            modifiedTriggerOrders.safeSet("stopLossOrder", modifiedStopLossOrder)
+            modifiedTriggerOrders.safeSet("takeProfitOrder", modifiedTakeProfitOrder)
+
+            modified["triggerOrders"] = modifiedTriggerOrders
+            modified
+        } else {
+            state
+        }
+    }
+
+    private fun calculateTriggerOrderTrade(
+        triggerOrder: Map<String, Any>,
+        position: Map<String, Any>,
+    ): Map<String, Any> {
+        val modified = calculatePrice(triggerOrder, position)
+        return modified
+    }
+
+    private fun calculatePrice(
+        triggerOrder: Map<String, Any>,
+        position: Map<String, Any>,
+    ): MutableMap<String, Any> {
+        val modified = triggerOrder.mutable()
+        val entryPrice = parser.asDouble(parser.value(position, "entryPrice.current"))
+        val inputType = parser.asString(parser.value(modified, "price.input"))
+
+        when (inputType) {
+            "stopLossOrder.price.triggerPrice" -> {
+                val triggerPrice = parser.asDouble(parser.value(modified, "price.triggerPrice"))
```

Comment:
nit: will it be any cleaner if we first parse all the triggerPirce / usdcDifff / percentDiff outside of this block?